les/peer.go

```diff
@@ -685,242 +896,278 @@ func (p *peer) Handshake(td *big.Int, head common.Hash, headNum uint64, genesis
-
-		recv.get("checkpoint/value", &p.checkpoint)
-		recv.get("checkpoint/registerHeight", &p.checkpointNumber)
-
-		if !p.onlyAnnounce {
-			for msgCode := range reqAvgTimeCost {
-				if p.fcCosts[msgCode] == nil {
-					return errResp(ErrUselessPeer, "peer does not support message %d", msgCode)
-				}
-			}
-		}
-		p.server = true
-	}
-	p.headInfo = &announceData{Td: rTd, Hash: rHash, Number: rNum}
-	return nil
-}
-
-// updateFlowControl updates the flow control parameters belonging to the server
-// node if the announced key/value set contains relevant fields
-func (p *peer) updateFlowControl(update keyValueMap) {
-	if p.fcServer == nil {
-		return
-	}
-	// If any of the flow control params is nil, refuse to update.
-	var params flowcontrol.ServerParams
-	if update.get("flowControl/BL", &params.BufLimit) == nil && update.get("flowControl/MRR", &params.MinRecharge) == nil {
-		// todo can light client set a minimal acceptable flow control params?
-		p.fcParams = params
-		p.fcServer.UpdateParams(params)
-	}
-	var MRC RequestCostList
-	if update.get("flowControl/MRC", &MRC) == nil {
-		costUpdate := MRC.decode(ProtocolLengths[uint(p.version)])
-		for code, cost := range costUpdate {
-			p.fcCosts[code] = cost
-		}
-	}
+		return nil
+	})
 }
 
-// String implements fmt.Stringer.
-func (p *peer) String() string {
-	return fmt.Sprintf("Peer %s [%s]", p.id,
-		fmt.Sprintf("les/%d", p.version),
-	)
+// serverPeerSubscriber is a callback interface to notify services about added or
+// removed server peers
+type serverPeerSubscriber interface {
+	registerPeer(*serverPeer)
+	unregisterPeer(*serverPeer)
 }
 
-// peerSetNotify is a callback interface to notify services about added or
-// removed peers
-type peerSetNotify interface {
-	registerPeer(*peer)
-	unregisterPeer(*peer)
+// clientPeerSubscriber is a callback interface to notify services about added or
+// removed client peers
+type clientPeerSubscriber interface {
+	registerPeer(*clientPeer)
+	unregisterPeer(*clientPeer)
 }
 
 // peerSet represents the collection of active peers currently participating in
 // the Light Ethereum sub-protocol.
 type peerSet struct {
-	peers      map[string]*peer
-	lock       sync.RWMutex
-	notifyList []peerSetNotify
-	closed     bool
+	serverPeers map[string]*serverPeer
+	clientPeers map[string]*clientPeer
+
+	// sSubs is a batch of subscribers and peerset will notify these
+	// subscribers when the peerset changes(new server peer is added
+	// or removed)
+	sSubs []serverPeerSubscriber
+
+	// cSubs is a batch of subscribers and peerset will notify these
+	// subscribers when the peerset changes(new client peer is added
+	// or removed)
+	cSubs []clientPeerSubscriber
+
+	closed bool
+	client bool
+	lock   sync.RWMutex
 }
 
 // newPeerSet creates a new peer set to track the active participants.
-func newPeerSet() *peerSet {
-	return &peerSet{
-		peers: make(map[string]*peer),
+func newPeerSet(client bool) *peerSet {
```

Comment:
If you insist on using a common `peerSet` then at least use separate constructors.